File: Algorithms/0377_Combination_Sum_IV/Python/Combination_Sum_IV_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def combinationSum4(self, nums, target):
        cache = {}

        def dfs(nums, target):
            if target < 0: return 0
            if target == 0: return 1
            if cache.get(target, -1) != -1: return cache[target]

            count = 0
            for num in nums:
                temp_target = target - num
                count += dfs(nums, temp_target)
            cache[target] = count

            return count

        dfs(nums, target)

        return cache[target]

File: Algorithms/0377_Combination_Sum_IV/Java/CombinationSumIvSolution1.java

// Space: O(n)
// Time: O(n^2)

public class Solution {
    public Map<Integer, Integer> cache = new HashMap<>();

    public int combinationSum4(int[] nums, int target) {
        dfs(nums, target);
        return cache.get(target);
    }

    private int dfs(int[] nums, int target) {
        if (target < 0) return 0;
        if (target == 0) return 1;
        if (cache.containsKey(target)) return cache.get(target);

        int count = 0;
        for (int num : nums) {
            int tempTarget = target - num;
            count += dfs(nums, tempTarget);
        }
        cache.put(target, count);
        return count;
    }
}




